Passive Crown Fire development

jkshuman opened the pull request:
Add equation to evaluate potential for crown fire at level of cohort and PFT
Add equation for passive crown fire ignition
Set fraction of crown burnt based on this conditional
Add crown fire threshold, cohort crown fire flag, and crown fire PFT param

<!--- Provide a general summary of your changes in the Title above -->

### Description:
This PR addresses passive crown fire development as a function of the stand. Within SPITFIRE crown fire ignition potential is evaluated against fire intensity using equations and logic from Bessie and Johnson (1995) that are drawn from Van Wagner (1977) at the stand level per Scott and Reinhardt (2001). Here the heat of crown foliage ignition is set to a constant value of 3060 kJ/kg based on 100% moisture (dry mass), but can be calculated via FATES-Hydro with additional development. (Comments indicate where this calculation would occur and be used this.) With successful passive crown fire ignition, active crown fire initiation is evaluated with PR #584 

Crown damage is evaluated via scorch height per usual.

### Collaborators:
main author: @jkshuman 
@slevisconsulting @lmkueppers @ckoven @pollybuotte @rgknox @rosiealice @xuchongang 

### Expectation of Answer Changes:
yes. this will change answers in fire runs that use crown fire PFTs. crown damage fraction has higher potential of being 100%.


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] My change requires a change to the documentation.
- [ ] I have updated the in-code documentation .AND. (the [technical note](https://github.com/NGEET/fates-docs) .OR. the wiki) accordingly.
- [ ] I have read the [**CONTRIBUTING**](https://github.com/NGEET/fates/blob/master/CONTRIBUTING.md) document.
- [ ] FATES PASS/FAIL regression tests were run
- [ ] If answers were expected to change, evaluation was performed and provided


### Test Results:
<!--- Non-trivial changes require the PASS/FAIL regression tests. -->
<!--- If changes to code are NOT expected to change answers, tests must -->
<!--- be run against a baseline. -->

CTSM (or) E3SM (specify which) test hash-tag:

CTSM (or) E3SM (specify which) baseline hash-tag:

FATES baseline hash-tag:

Test Output:

<!--- paste in test results here -->


<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

jkshuman commented:
I am still testing these changes, but welcome comments on the logic and flow of the changes.
@lmkueppers I had a call with @rgknox and @slevisconsulting and told them I would add this piece of code.

jkshuman commented on fire/SFMainMod.F90:
```diff
@@ -916,21 +922,53 @@ subroutine  crown_damage ( currentSite )
+                      
+                      ! Initiation of passive crown fire, EQ 9 Bessie and Johnson 1995                     
+                      currentCohort%passive_crown_fire = currentPatch%FI/currentCohort%crown_FI
+                      if (currentCohort%passive_crown_fire > 1.0_r8) then
```
that's a mistake. will update, thanks

jkshuman commented on parameter_files/fates_params_default.cdl:
```diff
@@ -131,7 +131,10 @@ variables:
+		fates_fire_crown_kill:long_name = "resistance to fire (1 = none), EQ 22 in Thonicke et al 2010" ;
+	double fates_fire_crown_fire(fates_pft) ;
+		fates_fire_crown_fire:units = "logical flag" ;
+		fates_fire_crown_fire:long_name = "Binary flag for passive crown fire";
```
@pollybuotte updated this per your comment.

jkshuman commented on fire/SFMainMod.F90:
```diff
@@ -940,8 +940,8 @@ subroutine  crown_damage ( currentSite )
                       ! Initiation of passive crown fire, EQ 9 Bessie and Johnson 1995                     
-                      currentCohort%passive_crown_fire = currentPatch%FI/currentCohort%crown_FI
-                      if (currentCohort%passive_crown_fire > 1.0_r8) then
+                      currentCohort%ignite_crown = currentPatch%FI/currentCohort%crown_FI
```
@pollybuotte does this make more sense?

slevisconsulting replied:
@jkshuman if the logic seems confusing, I would be fine with removing all the `else`s because you have already initialized `crown_fire = 0` and `fraction_crown_burned = 0` so they will end up with zero if none of the `if`s are true.

pollybuotte replied:
@jkshuman yes.

jkshuman replied:
@slevisconsulting I removed the extra detail, but retained the endif in latest commit

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -916,21 +922,53 @@ subroutine  crown_damage ( currentSite )
+                height_cbb = currentCohort%hite*EDPftvarcon_inst%crown(currentCohort%pft)
+                
+                ! Evaluate for passive crown fire ignition
+                if (EDPftvarcon_inst%crown_fire(currentCohort%pft) == 1) then
```
@jkshuman correct me if I'm mistaken:
@pollybuotte these are two different variables, and we should probably rename one of them.
1) `EDPftvarcon_inst%crown_fire` is the parameter that I added to the FATES params file for users to decide whether a pft CAN or CANNOT experience crown fire.
2) `currentCohort%crown_fire` is a variable that Jackie added for pfts that the user decided CAN experience crown fire. For those pfts, this variable switches to 1 if the model determines that conditions allow for crown fire.

pollybuotte replied:
OK, that makes sense. Renaming would help.

jkshuman replied:
@slevisconsulting @pollybuotte  I updated the currentCohort%crown_fire to be currentCohort%crown_fire_flag

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -916,21 +922,53 @@ subroutine  crown_damage ( currentSite )
+                   currentCohort%crown_fire = 0 ! not crown fire plant
+                   currentCohort%fraction_crown_burned =  0.0_r8
+                endif
+                
```
@jkshuman and everyone, I want to clarify that this way crown fire behavior is separate from scorch height. The canopy-bottom height enters the empirical formula for crown fuel ignition potential above, but scorch height does not influence the outcome. Is this ok?

jkshuman replied:
@slevisconsulting the scorch height (Patch%SH) is calculated in relation to the fire intensity (Patch%FI). 
https://github.com/jkshuman/fates/blob/f3077dfda1ceba060cb8493bf868835d8bfe7cd9/fire/SFMainMod.F90#L885-L886
currentPatch%SH = currentPatch%SH + f_ag_bmass * &
               EDPftvarcon_inst%fire_alpha_SH(currentCohort%pft) * (currentPatch%FI**0.667_r8) 

passive crown fire also uses the (Patch%FI) to evaluate if the energy of the fire is enough to ignite the crown. So, they are certainly connected.

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -916,21 +922,53 @@ subroutine  crown_damage ( currentSite )
+                      if (currentCohort%ignite_crown > 1.0_r8) then
+                         currentCohort%crown_fire_flg = 1  ! passive crown fire ignited
+                         currentCohort%fraction_crown_burned =  1.0_r8
+                      else ! evaluate crown damage based on scorch height
```
```suggestion

```

slevisconsulting replied:
...all the way to just before end if (line 957)

jkshuman replied:
@slevisconsulting I made these updates with last commit

slevisconsulting commented:
@jkshuman your changes seem good to me.

jkshuman commented:
This implementation addresses issue #567 and replaces the branch that @slevisconsulting started on that issue. The active development branch for this is https://github.com/jkshuman/fates/tree/passive_crown_fire

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -916,21 +923,49 @@ subroutine  crown_damage ( currentSite )
+                      ! Initiation of passive crown fire, EQ 14 Bessie and Johnson 1995
+                      currentCohort%ignite_crown = currentPatch%FI/currentCohort%passive_crown_FI
+                      
+                      if (currentCohort%ignite_crown > 1.0_r8) then
```
Minor correction. Looking at the article again I realized that this should be >=

```suggestion
                      if (currentCohort%ignite_crown >= 1.0_r8) then
```

This is the sentence in the paper that explains:
![image](https://user-images.githubusercontent.com/35322900/64462569-be303880-d0b5-11e9-8626-804e9add8373.png)

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -916,21 +923,49 @@ subroutine  crown_damage ( currentSite )
+                
+                ! Calculate clear branch bole height at base of crown
+                ! inst%crown = crown_depth_frac (PFT)
+                height_cbb = currentCohort%hite*EDPftvarcon_inst%crown(currentCohort%pft)
```
I am rethinking this one... To represent crown base height I think you need:
`crown_height - crown_depth`
which can be written as
`crown_height * (1 - crown_depth_frac)`

```suggestion
                height_cbb = currentCohort%hite * (1._r8 - EDPftvarcon_inst%crown(currentCohort%pft))
```

slevisconsulting replied:
So I think that your statement is calculating crown_depth rather than height_cbb:
`height_cbb = currentCohort%hite*EDPftvarcon_inst%crown(currentCohort%pft)`
 This series of suggestions that I've posted assumes that by crown base height we mean the height of the bottom of the canopy. I hope I have not misunderstood.

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -916,21 +923,49 @@ subroutine  crown_damage ( currentSite )
-                ! c%hite is height of cohort
-                if (currentPatch%SH < (currentCohort%hite-currentCohort%hite*EDPftvarcon_inst%crown(currentCohort%pft))) then 
+                ! c%hite is height of cohort  
+                if (currentPatch%SH < (currentCohort%hite- height_cbb)) .and. currentCohort%crown_fire_flg = 0 then 
```
If I'm correct in my last suggestion, then this becomes...

```suggestion
                if (currentPatch%SH < height_cbb .and. currentCohort%crown_fire_flg = 0) then 
```

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -916,21 +923,49 @@ subroutine  crown_damage ( currentSite )
                    ! flames over bottom of canopy but not over top.
                    if ((currentCohort%hite > 0.0_r8).and.(currentPatch%SH >=  &
-                        (currentCohort%hite-currentCohort%hite*EDPftvarcon_inst%crown(currentCohort%pft)))) then 
+                      (currentCohort%hite - height_cbb))) .and. currentCohort%crown_fire_flg = 0 then 
```
Following the same train of thought:

```suggestion
                      height_cbb) .and. currentCohort%crown_fire_flg = 0) then 
```

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -916,21 +923,49 @@ subroutine  crown_damage ( currentSite )
-                           currentCohort%fraction_crown_burned =  (currentPatch%SH-currentCohort%hite*(1.0_r8- &
-                                EDPftvarcon_inst%crown(currentCohort%pft)))/(currentCohort%hite* &
-                                EDPftvarcon_inst%crown(currentCohort%pft)) 
+                      currentCohort%fraction_crown_burned =  (currentPatch%SH-currentCohort%hite*(1.0_r8- &
```
...so in this line you'd say:

```suggestion
                      currentCohort%fraction_crown_burned =  (currentPatch%SH - height_cbb) &
```

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -916,21 +923,49 @@ subroutine  crown_damage ( currentSite )
-                                EDPftvarcon_inst%crown(currentCohort%pft)))/(currentCohort%hite* &
-                                EDPftvarcon_inst%crown(currentCohort%pft)) 
+                      currentCohort%fraction_crown_burned =  (currentPatch%SH-currentCohort%hite*(1.0_r8- &
+                             EDPftvarcon_inst%crown(currentCohort%pft)))/(height_cbb) 
```
...while in this line you'd say:

```suggestion
                             / (currentCohort%hite * EDPftvarcon_inst%crown(currentCohort%pft))
```

slevisconsulting replied:
@jkshuman pls let me know if anything I'm suggesting is unclear and also if you think that I have misunderstood.

jkshuman replied:
@slevisconsulting Thanks for checking this over. I agree with your assessment, and I fixed this based on your suggestions. Please let me know if you think it looks good now.

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -928,9 +928,10 @@ subroutine  crown_damage ( currentSite )
                 ! inst%crown = crown_depth_frac (PFT)
-                height_cbb = currentCohort%hite*EDPftvarcon_inst%crown(currentCohort%pft)
+                height_cbb  = currentCohort%hite * (1.0_r8 - EDPftvarcon_inst%crown(currentCohort%pft)
+                crown_depth = currentCohort%hite*EDPftvarcon_inst%crown(currentCohort%pft) 
```
@jkshuman this looks correct to me. 
A nit-picky request would be to reorder lines 933 and 934 so as to rewrite height_cbb as:

`height_cbb = currentCohort%hite - crown_depth`

jkshuman replied:
@slevisconsulting I like it. That's what I was going for in the first place, but this makes it super clear. done.

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -954,18 +955,18 @@ subroutine  crown_damage ( currentSite )
 
-                      currentCohort%fraction_crown_burned =  (currentPatch%SH-currentCohort%hite*(1.0_r8- &
-                             EDPftvarcon_inst%crown(currentCohort%pft)))/(height_cbb) 
+                      currentCohort%fraction_crown_burned =  (currentPatch%SH - height_cbb)/crown_depth 
```
@jkshuman I'm realizing that I prefer the following alternative here:
1) remove line 966:
`.and. (currentPatch%SH <= currentCohort%hite)  &`
2) change line 969 to:
`currentCohort%fraction_crown_burned =  max(0.0_r8, min(1.0_r8, (currentPatch%SH - height_cbb)/crown_depth))`
3) remove the `else  ! Flames over top of canopy.` lines 971-973

The advantage of this is that you make lines 971-973 unnecessary and maintain the relevance of
`currentCohort%hite > 0.0_r8`
`currentPatch%SH > height_cbb`
`currentCohort%crown_fire_flg = 0`
that currently appear in the if statement but not in the else.

So leaving the code as is, you could end up with 
`currentCohort%fraction_crown_burned =  1.0_r8`
for `currentCohort%hite = 0.0`  (probably not a big deal)
or for `currentCohort%crown_fire_flg = 1`  (not a problem at the moment, but...)
The latter would become a problem if we decided to allow passive crown fires to give
`currentCohort%fraction_crown_burned < 1.0_r8`
because then the else would overwrite the value with 1.0_r8.

Last thing...
If you do change these lines as I suggest, then you could clean up further by moving the `if` in lines 965-967 to line 961 and make into an `else if`

Let me know if I'm making sense of if I've confused myself and am completely wrong about this.

jkshuman replied:
@slevisconsulting I agree with you on these. made some updates.

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -916,28 +926,52 @@ subroutine  crown_damage ( currentSite )
+                
+                ! Flames lower than bottom of canopy.
+                ! height_cbb is clear branch bole height or height of bottom of canopy 
+                if (currentPatch%SH <= height_cbb) .and. (currentCohort%crown_fire_flg == 0) then &
```
@jkshuman this was fine like this, i.e. without the parentheses in the middle near the `.and.` and without the `&` at the end:
```suggestion
                if (currentPatch%SH <= height_cbb .and. currentCohort%crown_fire_flg == 0) then
```
If you prefer more parentheses, you could do 
```suggestion
                if ((currentPatch%SH <= height_cbb) .and. (currentCohort%crown_fire_flg == 0)) then
```

jkshuman replied:
@slevisconsulting Yeah, I was hitting a wall debugging this section, and missed the incorrect &. the parentheses got thrown in there as I was working through things. It's running and compiling now, so maybe I will remove the extra parentheses...

slevisconsulting replied:
Now I realize that we don't even need this `if` in the logic because you initialized fraction_crown_burned = 0.0_r8 to begin with.  So we can simply turn the next `else if` into an `if`.  (Sorry for catching these one-at-a-time...)

slevisconsulting replied:
> [...] It's running and compiling now [...]

That's awesome! I'll wait for a "go ahead" from you, to update my `active_crown_fire` branch. Then I will create a PR to replace the accidental one that needs to close.

jkshuman replied:
@slevisconsulting good point. I added some comments for clarity around surface fires, and updated this.

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -916,28 +926,52 @@ subroutine  crown_damage ( currentSite )
+                ! Flames part way into canopy
+                ! Equation 17 in Thonicke et al. 2010
+                elseif ((currentCohort%hite > 0.0_r8).and.(currentPatch%SH > height_cbb) &
+                     .and. (currentCohort%crown_fire_flg == 0) then  &
```
Similar here. Remove the `&` at the end and add one more parenthesis at the end for balance:
```suggestion
                     .and. (currentCohort%crown_fire_flg == 0)) then
```

slevisconsulting commented on main/EDTypesMod.F90:
```diff
@@ -146,6 +146,7 @@ module EDTypesMod
   integer,  parameter, public :: lg_sf                = 6          ! array index of live grass pool for spitfire
 
   real(r8), parameter, public :: fire_threshold       = 50.0_r8    ! threshold for fires that spread or go out. KWm-2 (Pyne 1986)
+  real(r8), parameter, public :: crown_fire_threshold = 200.0_r8   ! threshold for passive crown fire ignition. KWm-2 (Bessie & Johnson 1995)
```
Just noticed that this parameter appears here, as well as in SFMainMod line 917.

jkshuman replied:
@slevisconsulting yep. that was my debugging frustration. corrected this by bringing variable into SFMain from EDTypes near top and removing line 917 in SFMain

jkshuman commented:
@slevisconsulting doing a bit of final cleanup, but nearly "done"
everything compiles and runs for 40+ years with active fire and constant ignitions across South America.

pollybuotte commented:
@jkshuman There was a problem with my parameter file.  Just had a successful one year run so will test longer now.

jkshuman commented:
Yep. I had a similar issue. Glad you figured it out.

On Fri, Sep 13, 2019 at 9:29 AM pollybuotte <notifications@github.com>
wrote:

> @jkshuman <https://github.com/jkshuman> There was a problem with my
> parameter file. Just had a successful one year run so will test longer now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NGEET/fates/pull/572?email_source=notifications&email_token=AFIUHBTJGLI4Z67QOAVHO33QJOWUHA5CNFSM4ITXDGNKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6VL4GY#issuecomment-531283483>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFIUHBU3YF44KVXHKXMAC5DQJOWUHANCNFSM4ITXDGNA>
> .
>
-- 
Jacquelyn Shuman
Terrestrial Sciences Section
NCAR

slevisconsulting commented:
I made two crown-related history variables _active_ in my branch (https://github.com/jkshuman/fates/pull/5) and then tested over the "transect" domain by running from bare ground for 20 yrs without fire and then 20 years with fire.

What's strange is that fire occurs in the first year of the second run and does not occur again after that. Did this happen in your CZ2 run @pollybuotte ?

pollybuotte commented:
My CZ2 run fails on July 24 of year 2 with errors in the cesm.log that indicate problems with fire_litter_fluxes.

0:MPT:     currentsite=..., 
0:MPT:     currentpatch=<error reading variable: value requires 163904 bytes, which is more than max-value-size>, 
0:MPT:     newpatch=<error reading variable: value requires 163904 bytes, which is more than max-value-size>, patch_site_areadis=9.3355233037141527)
0:MPT:     at /glade/u/home/pbuotte/git/passive_crown/src/fates/biogeochem/EDPatchDynamicsMod.F90:1497
0:MPT: #7  0x000000000135d78f in edpatchdynamicsmod::spawn_patches (currentsite=..., 
0:MPT:     bc_in=...)
0:MPT:     at /glade/u/home/pbuotte/git/passive_crown/src/fates/biogeochem/EDPatchDynamicsMod.F90:561
0:MPT: #8  0x000000000134366d in edmainmod::ed_ecosystem_dynamics (currentsite=..., 
0:MPT:     bc_in=...)
0:MPT:     at /glade/u/home/pbuotte/git/passive_crown/src/fates/main/EDMainMod.F90:241
0:MPT: #9  0x00000000008eb0b6 in clmfatesinterfacemod::dynamics_driv (
0:MPT:     this=0x56334c0 <clm_instmod_mp_clm_fates_>, nc=1, bounds_clump=..., 
0:MPT:     atm2lnd_inst=..., soilstate_inst=..., temperature_inst=..., 
0:MPT:     waterstate_inst=..., canopystate_inst=..., 
0:MPT:     soilbiogeochem_carbonflux_inst=..., frictionvel_inst=...)
0:MPT:     at /glade/u/home/pbuotte/git/passive_crown/src/utils/clmfates_interfaceMod.F90:681
0:MPT: #10 0x00000000008925cc in clm_driver::clm_drv (doalb=4294967295, 
0:MPT:     nextsw_cday=205.02083333333334, declinp1=0.35031168169655491, 
0:MPT:     declin=0.35031168169655491, rstwr=.FALSE., nlend=.FALSE., rdate=..., 
0:MPT:     rof_prognostic=.FALSE., .tmp.RDATE.len_V$20ab2=32)
0:MPT:     at /glade/u/home/pbuotte/git/passive_crown/src/main/clm_driver.F90:879
0:MPT: #11 0x000000000084436b in lnd_comp_mct::lnd_run_mct (eclock=..., cdata_l=..., 
0:MPT:     x2l_l=..., l2x_l=...)
0:MPT:     at /glade/u/home/pbuotte/git/passive_crown/src/cpl/lnd_comp_mct.F90:451
0:MPT: #12 0x000000000045ed1c in component_mod::component_run (eclock=..., comp=..., 
0:MPT:     infodata=..., seq_flds_x2c_fluxes=..., seq_flds_c2x_fluxes=..., 
0:MPT:     comp_prognostic=4294967295, comp_num=2, timer_barrier=..., 
0:MPT:     timer_comp_run=..., run_barriers=.FALSE., ymd=19800724, tod=1800, 
0:MPT:     comp_layout=..., .tmp.SEQ_FLDS_X2C_FLUXES.len_V$341f=4096, 
0:MPT:     .tmp.SEQ_FLDS_C2X_FLUXES.len_V$3422=4096, 
0:MPT:     .tmp.TIMER_BARRIER.len_V$3427=19, .tmp.TIMER_COMP_RUN.len_V$342a=11, 
0:MPT:     .tmp.COMP_LAYOUT.len_V$3430=32)
0:MPT:     at /glade/u/home/pbuotte/git/passive_crown/cime/src/drivers/mct/main/component_mod.F90:724
0:MPT: #13 0x000000000042a027 in cime_comp_mod::cime_run ()
0:MPT:     at /glade/u/home/pbuotte/git/passive_crown/cime/src/drivers/mct/main/cime_comp_mod.F90:2447
0:MPT: #14 0x000000000044767c in cime_driver ()
0:MPT:     at /glade/u/home/pbuotte/git/passive_crown/cime/src/drivers/mct/main/cime_driver.F90:133
0:MPT: #15 0x0000000000407f1e in main ()
0:MPT: #16 0x00002b0972dbe6e5 in __libc_start_main ()
0:MPT:    from /glade/u/apps/ch/os/lib64/libc.so.6
0:MPT: #17 0x0000000000407e29 in _start () at ../sysdeps/x86_64/start.S:118
0:MPT: (gdb) A debugging session is active.
0:MPT: 
0:MPT:  Inferior 1 [process 37486] will be detached.

Log files are here:
/glade/scratch/pbuotte/CZ2_passive_crown/run

jkshuman commented:
Updated this to have crown_ignite_energy be a PFT level variable. This will allow for easier integration with FATES-Hydro as referenced in comments for active crown fire 
https://github.com/NGEET/fates/issues/573
tagging @xuchongang

jkshuman commented:
@pollybuotte and @slevisconsulting @lmkueppers I tested the functionality of passive crown fire in runs across the tropics, but it will need to be tested and calibrated for CA.

jkshuman commented:
@slevisconsulting @pollybuotte I should note that I see similar behavior with this junk test of passive crown fire (junk because it is a default uncalibrated test param file run across South America). There is active fire for the first 5 years, and then no subsequent fire area. There is still vegetation, suggesting that the fuels have been removed and there is not enough energy to trigger fire. This tells me that local calibration and ability to capture surface fuels is essential to this crown fire regime. Probably should put in a write that indicates if crown fire was triggered, and then look at the state of the fuels.

rgknox commented on fire/SFMainMod.F90:
```diff
@@ -902,28 +911,56 @@ subroutine  crown_damage ( currentSite )
+                      
+                      if (ignite_crown >= 1.0_r8) then
+                         currentCohort%crown_fire_flg = 1  ! passive crown fire ignited
+                         currentCohort%fraction_crown_burned =  1.0_r8
```
initialized in the case that this logic is never triggered, check

jkshuman replied:
@rgknox can you expand on where you want this check?

jkshuman replied:
@rgknox thanks for confirming this, and that no updates are needed.

rgknox commented on fire/SFMainMod.F90:
```diff
@@ -902,28 +911,56 @@ subroutine  crown_damage ( currentSite )
+                if (currentCohort%hite > 0.0_r8 .and. currentPatch%SH > height_cbb &
+                     .and. currentCohort%crown_fire_flg == 0) then  
+
+                      currentCohort%fraction_crown_burned = max(0.0_r8, &
```
Since line 965 places the constraints between 0 and 1, it seems redundant here.

rgknox commented on fire/SFMainMod.F90:
```diff
@@ -962,7 +999,8 @@ subroutine  cambial_damage_kill ( currentSite )
           currentCohort => currentPatch%tallest;
-          do while(associated(currentCohort))  
+          do while(associated(currentCohort))
+             currentCohort%cambial_mort = 0.0_r8
```
nice, good place to zero

jkshuman replied:
on board with the clean up of zeroing...

rgknox commented on main/EDPftvarcon.F90:
```diff
@@ -48,9 +48,11 @@ module EDPftvarcon
                                                      ! that is occupied by crown. For fire model. 
      real(r8), allocatable :: bark_scaler(:)         ! scaler from dbh to bark thickness. For fire model.
-     real(r8), allocatable :: crown_kill(:)          ! scaler on fire death. For fire model. 
+     real(r8), allocatable :: crown_resist(:)        ! resistance to fire. (1 = none) For fire model.
```
I feel this naming convention is un-intuitive.  If the value is increasing, one would think the resistance is increasing, but its the opposite.

jkshuman replied:
yep, good point. rolling back to crown_kill and updated comments

rgknox commented on main/EDPftvarcon.F90:
```diff
@@ -48,9 +48,11 @@ module EDPftvarcon
      real(r8), allocatable :: bark_scaler(:)         ! scaler from dbh to bark thickness. For fire model.
-     real(r8), allocatable :: crown_kill(:)          ! scaler on fire death. For fire model. 
+     real(r8), allocatable :: crown_resist(:)        ! resistance to fire. (1 = none) For fire model.
+     real(r8), allocatable :: crown_fire(:)          ! Does plant have passive crown fire? (1=yes, 0=no)
```
Perhaps change "have" to "susceptible"?

rgknox replied:
also, maybe this parameter should have the name passive in it?

jkshuman replied:
updated name and comments

rgknox commented on main/EDPftvarcon.F90:
```diff
@@ -377,7 +379,15 @@ subroutine Register_PFT(this, fates_params)
+    call fates_params%RegisterParameter(name=name, dimension_shape=dimension_shape_1d, &
+         dimension_names=dim_names, lower_bounds=dim_lower_bound)
+
+    name = 'fates_fire_crown_fire'
```
almost a palindrome.

rgknox commented on fire/SFMainMod.F90:
```diff
@@ -902,28 +911,56 @@ subroutine  crown_damage ( currentSite )
+                height_cbb                    = currentCohort%hite - crown_depth
+                passive_crown_FI              = 0.0_r8  !critical fire intensity for passive crown fire
+                ignite_crown                  = 0.0_r8  !ratio for ignition of passive crown fire,EQ 14 Bessie & Johnson 1995
+                currentCohort%crown_fire_flg  = 0       !flag for passvie crown fire ignition
```
as far as I can tell, crown_fire_flg is only used inside this cohort loop. 
Can we turn this into a local variable? Also, I would recommend using a logical, and also having the name passive in there somewhere.

jkshuman replied:
@rgknox updated this flag to logical. I think when I wrote this originally, I left it at the cohort level so we can collect history output on this (passive or active crown fire mortality cause). But I never added that history output. Might check in with you on that history logic/format.

jkshuman replied:
@rgknox per our email yesterday, this was reverted to binary (1 = true).

jkshuman commented:
@rgknox thanks for the comments. I will implement them this week. Just a note that this PR #572 and #573 will both have changes to the parameter file, so I think it would be good to wait and implement them together. Reviewing now is helpful though, as @slevisconsulting can incorporate these updates into #573 .

rgknox commented:
Ok, thanks for info @jkshuman .  We may also want to see if others have new parameters to change in the near future, so we can also lump this all into 1 change.  

For instance, @rosiealice , introduced a new parameter that sets the minimum storage allowed to be left after leaf flushing, maybe we could add that to the group.

For nutrients, I know we will soon need at least  vmax and the michaelis menten pft parameters for acquisition, I think it would be good to add these as well.

jkshuman commented:
@rgknox I addressed your review comments. I did not add the crown_fire flag to history output. But left it as a cohort attribute so easy to add when there is time/interest.
@slevisconsulting please fetch these updates for #573 
@pollybuotte @slevisconsulting  be aware there was a change to the param file based on review by @rgknox

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -902,28 +911,55 @@ subroutine  crown_damage ( currentSite )
+                   ! 0.01 = C from Van Wagner 1977 EQ4 for crown of base height 6m, 100% FMC, and FI 2500kW/m
+                   passive_crown_FI = (0.01_r8 * height_cbb *EDPftvarcon_inst%crown_ignite_energy(currentCohort%pft))**1.5_r8
+
+                   if (passive_crown_FI >= crown_fire_threshold) then ! 200 kW/m = threshold for crown fire potential
```
@jkshuman , we (@pollybuotte @lmkueppers and myself) were looking more carefully at this if statement, and we convinced ourselves that it should change to:
```suggestion
                   if (currentPatch%FI >= crown_fire_threshold) then ! 200 kW/m = threshold for crown fire potential
```
because as written it says that `height_cbb = 1` is less likely to pass the if test than a `height_cbb = 10` which is the opposite of what should happen. Similarly if you applied a moisture dependence (as you say in line 933), a smaller m would be less likely to pass the if test than a larger m, which again is the opposite of what should happen.

jkshuman replied:
Ah yes. Good catch!

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -902,28 +911,55 @@ subroutine  crown_damage ( currentSite )
-                endif
+                
+                ! Evaluate for passive crown fire ignition
+                if (EDPftvarcon_inst%crown_fire(currentCohort%pft) == 1) then
```
@jkshuman in discussions with @lmkueppers and @pollybuotte we decided to change this line to:
```suggestion
                if (EDPftvarcon_inst%crown_fire(currentCohort%pft) > 0.0_r8) then
```
and change `crown_fire` from integer to real. This way we can rank trees by their vulnerability to passive crown fire in a non-binary way. I will add a comment below to show how we envision doing this.

slevisconsulting commented on fire/SFMainMod.F90:
```diff
@@ -902,28 +911,55 @@ subroutine  crown_damage ( currentSite )
+                      
+                      if (ignite_crown >= 1.0_r8) then
+                         currentCohort%passive_crown_fire_flg = 1 ! passive crown fire ignited
+                         currentCohort%fraction_crown_burned =  1.0_r8
```
@jkshuman we would change this line to:
```suggestion
                         currentCohort%fraction_crown_burned =  EDPftvarcon_inst%crown_fire(currentCohort%pft)
```
to account for pft vulnerability to passive crown fire, instead of setting to 1 for all pfts.

jkshuman replied:
@slevisconsulting sure. then users can decide how to implement

jkshuman commented:
@slevisconsulting @lmkueppers @pollybuotte I added your updates, as well as a comment from a ref that @adrifoster shared. There is potential survival from passive crown fire! This adds nuance to simulating this, as I think there will be situations where the tree should have 100% crown damage. Good to add the functionality to track this.

jkshuman commented:
Chatting with @adrifoster and per Scott and Reinhardt 2001 there is a difference between crown base height and canopy base height. Current implementation uses crown base height (clear branch bole height). Canopy base height is the lowest height  where canopy fuel is sufficient to propagate fire (0.011km/m3). @adrifoster has a method to get this value, and we will implement soon. Expectation is that this will not change passive crown fire, but has the capacity to change active crown dynamics (small tree cannot initiate an active crown fire).

rgknox commented on main/EDPftvarcon.F90:
```diff
@@ -49,8 +49,7 @@ module EDPftvarcon
      real(r8), allocatable :: bark_scaler(:)         ! scaler from dbh to bark thickness. For fire model.
      real(r8), allocatable :: crown_kill(:)          ! crown resistance to fire. (1 = none) For fire model.
-     real(r8), allocatable :: active_crown_fire(:)   ! Is plant susceptible to active crown fire?(1=yes,0=no)
-     real(r8), allocatable :: foliar_moisture(:)     ! foliar moisture content from dry fuel [%]
```
Strange, it wants to remove foliar_moisture, but foliar_moisture is not in master.. I wonder if git is confused...?

jkshuman commented:
Implementation EQ 14 of Bessie and Johnson has been updated per Scott and Reinhardt 2001 to consider stand level canopy fuel density versus tree level fuels for crown fire initiation on a patch #572 and integrated into #584 . edited in PR above. Canopy base height for the stand is the height at which a minimum  fuel bulk density of (0.011 kg/m3) is found. This accounts for variation in fuels within multi-story stands.

ckoven commented:
@jkshuman can we close this PR, since all the developments are in either #584 (not ready) or #594 (about to be merged)?

jkshuman commented:
closing this per @ckoven. All crown fire development has been merged into PR #584 and will continue development on that PR. Other pieces will be merged per #594.
